app/models/manageiq/providers/google/cloud_manager/event_catcher/stream.rb

```diff
@@ -0,0 +1,62 @@
+# Helper class responsible for polling a Google Pubsub topic and retreiving
+# event type messages. Parsing of the message is deferred to the caller of
+# #poll.
+class ManageIQ::Providers::Google::CloudManager::EventCatcher::Stream
+  def initialize(ems)
+    @ems = ems
+  end
+
+  # Start capturing events
+  def start
+    @collecting_events = true
+  end
+
+  # Stop capturing events
+  def stop
+    @collecting_events = false
+  end
+
+  # Poll for events (blocks forever until #stop is called)
+  def each_batch
+    while @collecting_events
+      yield events.map { |e| JSON.parse(e.message['data']) }
+    end
+  end
+
+  private
+
+  # Poll for messages. This makes a service call to the Google Pubsub service
+  # and returns any messages generated from the activity log. Note that these
+  # messages are acknowledged before returning, so the caller has no
+  # responsibility to explicitly acknowledge them.
+  #
+  # @return [Array<Fog::Google::Pubsub::ReceivedMessage>] possibly empty list of messages
+  def events
+    # For now, return immediately with up to 10 messages
+    @ems.with_provider_connection(:service => 'pubsub') do |google|
+      subscription = get_or_create_subscription(google)
+
+      msgs = subscription.pull(:return_immediately => true, :max_messages => 10)
+      subscription.acknowledge(msgs)
+
+      msgs
+    end
+  end
+
+  def get_or_create_subscription(google)
+    subscription = google.subscriptions.get(subscription_name)
```

Comment:
this could be even less verbose and more precise
```ruby
def get_or_create_subscription(google)
  google.subscriptions.tap do |sub|
     sub.get(subscription_name) || sub.create(:name  => subscription_name, :topic => topic_name)
  end
end